### DISCUSSION TITLE
Welcome to AITOOL Discussions!

### DISCUSSION THREAD

**Discussion by VorlonCD** [VorlonCD/bi-aidetection#216] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

I'm enabling this up as an alternative to the famous 50,000 page thread on ipcamtalk (that I tend to ignore):
https://ipcamtalk.com/threads/tool-tutorial-free-ai-person-detection-for-blue-iris.37330/

Regular bugs or feature requests should still be added over here:
https://github.com/VorlonCD/bi-aidetection/issues

Releases are here, but are often outdated.   I'm much more likely to post an updated version in an issues thread, right or wrong.
https://github.com/VorlonCD/bi-aidetection/releases

And as always, a huge thanks to @gentlepumpkin for starting this awesome little project!   Also thanks to @classObject for creating the masking features, and @Tinbum1 for suggesting a ton of cool features and putting up with my bugs along the way.

**Vorlon**

**Comment by balucanb** [comment]
Great Idea VorlonCD!

**Comment by johnjoemorgan** [comment]
I feel we should celebrate with some Vogon poetry! ;-)

**Reply by VorlonCD** [reply]
Blasphemy!   Vogon != Vorlon  :)

![image](https://user-images.githubusercontent.com/7512113/112632722-e0f87e00-8e0e-11eb-849c-e5c063fa0ada.png)

**Comment by balucanb** [comment]
Based on the speed and amount of changes made/feature requests granted and so on, the only viable conclusion is VorlonCD is apparently a robot and never requires sleep..or maybe he has a clone army working on this project. I digress. Seriously though I am finding it harder and harder to stay up on the changes and how said changes are effecting? / affecting? the operation of this wonderful project, is there a wiki, how to or explanation of all the workings anywhere? There should be!  Yes I am asking for it whilst never offering to help do it..I suck. Sorry. I really would like to help with the solution but I live directly in the "Clueless User" category.

**Comment by balucanb** [comment]
@VorlonCD or anyone else...pretty long story so I wanted to ask for some help here vs. putting in a bug report when it may not be a bug.
Not getting relevant hits like I think I should. trying to figure out if it is my settings or the program. 
Running version v2.0.1036
What I have done:
1. I disabled all my cameras except 1 which takes a snapshot ever 4 seconds.
2. All the settings on the camera page (Default, Telegram, Dynamic masking) have been reset to defaults (30%)
3. Dynamic Masking is enabled. only setting changed there were clear history is 60 min. Under percent match, Advanced, scaling is enabled and small objects is set to 2 / 90 Midsize objects set to 5/8/95.
4. Removed all masks from this camera.
5. Went thru the snapshot folder and found 29 pics that clearly should have been picked up, copied those, deleted everything in the folder. cleared everything in AI-Tool, shut it down, restarted it and dropped those 29 pics into the folder.
6. 4 of the 29 then shown up as relevant.
7. Here is one that clearly should have hit as relevant but did not, I think 95% of the ones that did not hit should have:
When you view it under history the heading over picture says 3x irrelevant: Car (78%); Car (53%); Car (39%)
If you check the prediction Object Detail tab there are 3 entries and under the confidence heading they are 77.6%, 52.6 and 39.4%
% of image entries are 0.8, 0.4 and 0.3%
Not sure if knowing the BI settings are needed so I will leave those out unless asked for. If you need anything else let me know and THANKS!
![rsyardsd 20210404_104508323_AITOOLTEST_43333 2253094](https://user-images.githubusercontent.com/34525580/113517335-ec515500-9544-11eb-993e-26fc4a7ac8e3.jpg)

**Reply by VorlonCD** [reply]
You can quickly re-test past history items by right-click history tab item > **Test Detection Again** rather than dropping images in the monitored folder.

If you test more than a few images with objects detected in the same area, Dynamic Masking will kick in - You can tell by viewing History tab > Prediction Details > **Result**, **DynMaskResult** and **ImgMaskResult** columns.    It should tell you if anything was masked.

I finally tested your previous telegram issue, and it seems to be getting past the initial relevant object check, then skipping telegram if the same object has a higher threshold in telegram relevant objects, so that seems to be working.   I did add better telegram logging when I was in there, and 'blurred' chat id and API key in the log to make it more private.

I also added better relevant object logging in this update - Look for "**RelevantObjectManager**" references in the log.
[AITool-VorlonCD-v2.0.1040.zip](https://github.com/VorlonCD/bi-aidetection/files/6255469/AITool-VorlonCD-v2.0.1040.zip)

I'll need screen shot of the "Prediction Details" screen for something that failed, or you can select all and ctrl-c the text into the clipboard.

And after testing, on the LOG tab hit PAUSE, and scroll up to find the log entries for the last image processed.   Those lines can be ctrl-c copied also.

**Reply by balucanb** [reply]
Thank you! I will let you know how it works.

**Reply by balucanb** [reply]
Getting the following with the latest, getting alerts to telegram but AI tool is showing error on the deepstack tab.


Date | Func | Detail | Level | Source | AIServer | Camera | Image | Idx | Depth | Color | ThreadID | FromFile | Filename
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
4/5/2021 06:13:06 | StartDeepstack | 1 'redis-server.exe' processes did not start within 189ms | Error | AITOOLS.EXE | None | None | None | 64 | 2 |   | 6 | False | AITool.[2021-04-05].log
4/5/2021 06:13:46 | StartDeepstack | 1 'redis-server.exe' processes did not start within 380ms | Error | AITOOLS.EXE | Deepstack_Objects:127.0.0.1:81 | RSyardsd | None | 146 | 5 |   | 9 | False | AITool.[2021-04-05].log
4/5/2021 06:13:47 | DetectObjects | The underlying connection was closed: An unexpected error occurred on a receive. [WebException] (In AITOOL.GetDetectionsFromAIServer:2045) | Error | AITOOLS.EXE | Deepstack_Objects:127.0.0.1:81 | RSyardsd | RSyardsd.20210405_061320841.jpg | 153 | 2 |   | 15 | False | AITool.[2021-04-05].log
4/5/2021 06:17:12 | ResetDeepstack | The process cannot access the file 'stderr.txt' because it is being used by another process. [IOException] (In DeepStack.ResetDeepstack:931) | Error | AITOOLS.EXE | Deepstack_Objects:127.0.0.1:81 | RSyardsd | None | 1716 | 4 |   | 1 | False | AITool.[2021-04-05].log
4/5/2021 06:17:14 | DetectObjects | Unable to connect to the remote server [WebException] (In AITOOL.GetDetectionsFromAIServer:2045) | Error | AITOOLS.EXE | Deepstack_Objects:127.0.0.1:81 | RSyardsd | RSyardsd.20210405_061712334.jpg | 1730 | 1 |   | 19 | False | AITool.[2021-04-05].log
4/5/2021 06:17:18 | DetectObjects | Unable to connect to the remote server [WebException] (In AITOOL.GetDetectionsFromAIServer:2045) | Error | AITOOLS.EXE | Deepstack_Objects:127.0.0.1:81 | RSyardsd | RSyardsd.20210405_061716322.jpg | 1763 | 1 |   | 20 | False | AITool.[2021-04-05].log
4/5/2021 06:17:19 | StartDeepstack | 1 'redis-server.exe' processes did not start within 305ms | Error | AITOOLS.EXE | Deepstack_Objects:127.0.0.1:81 | RSyardsd | None | 1789 | 2 |   | 16 | False | AITool.[2021-04-05].log

**Comment by VorlonCD** [comment]
hit the STDERR button on deepstack tab to see the error.   Try a reboot

**Reply by balucanb** [reply]
did that no help. Of course I forgot to do a B/U so I just deleted the contents of the folder, rebooted and installed the version previous but Deepstack will not start now? Any suggestions? No contents in the STDERR file BTW.

**Reply by balucanb** [reply]
Think I fixed it. something was wonky with python and the res server thing. I just killed everything in task manager and re started. seems to be ok now. 

EDIT-  Sucks that I had to redo everything since I had no back up, with the clean install I am back to the telegram issue. I only added the 1 camera back left all percentages at default, telegram going off like crazy, which I would expect. Went in and changed telegram to 60% for all vehicle types (everything else is still 30%) but it is still sending notifications for items less than 60%? Also Deepstack keeps erroring out, actually had some info in the log.
[stderr (2).txt](https://github.com/VorlonCD/bi-aidetection/files/6258880/stderr.2.txt)
[stdout.txt](https://github.com/VorlonCD/bi-aidetection/files/6258884/stdout.txt)
[AITool.[2021-04-05].log](https://github.com/VorlonCD/bi-aidetection/files/6258971/AITool.2021-04-05.log)

**Reply by VorlonCD** [reply]
@balucanb Will try to take a closer look at your log in a day or two.  

For future reference the latest build will always be here in the future and it has an **actual installer:**  (nothing changed for your issue yet)
https://github.com/VorlonCD/bi-aidetection/tree/master/src/UI/Installer

To find the actual deepstack crash error, backup this file and replace with my modified copy, and stop/start deepstack.   Then take a look at stderr.txt again after the next crash.    It prevents the premature crash in your last TXT file so the actual error can be displayed.

"C:\DeepStack\intelligencelayer\shared\detection.py"

(unblock the ZIp after downloading!)
[detection.zip](https://github.com/VorlonCD/bi-aidetection/files/6261601/detection.zip)

**Reply by VorlonCD** [reply]
@balucanb that AITOOL log doesnt really have any references about telegrams being sent.   See if you can post a new log from after you get a notification you think you should not.   Tell me the date/time it happened, and the camera name.

**Reply by balucanb** [reply]
Will do. I have added back a couple more cameras and other than some trees being id'ed as people the Telegram hits seem to be acting correct since I did the clean install...I will let you know it it gets uppity again.

**Reply by VorlonCD** [reply]
Remember that if you have Settings > Send Errors Telegram, THAT would be giving you notifications also if you were having deepstack errors.

**Comment by balucanb** [comment]
Ref. the installer package(s) @VorlonCD I went there and tried it but wondering if I did something wrong, it installed a complete new instance of AI Tool instead of just overwriting the files in the folder where I currently have it, I did  change the install location. Once it loaded, flawlessly BTW I saw where it was brand new and none of my cameras were there etc. I just uninstalled it and stayed with what I had. Did I miss a step? Will eventually use it I am sure but want to keep all my settings and so on. THANKS!

**Reply by VorlonCD** [reply]
The **`\_SETTINGS\`** subfolder contains all the camera settings, history entries and mask images, so you can just copy that manually to a new AITOOL folder if you want to preserve settings.   

The installer wont know the first time what the last install path was, but it should remember it after that.

**Comment by aesterling** [comment]
@VorlonCD out of curiosity, in what format does Deepstack return object data back to AI Tool? Is it just a long string of ugly text that AI Tool parses to present these nicely formatted variables? (Like objects in Title Case, and confidence percentages in whole numbers.)

Is it possible to see the "raw" data that DS returns in the AI Tool log? Thanks!

![variables_window](https://user-images.githubusercontent.com/18314707/115048673-6a93ec80-9e9f-11eb-9db7-b184f5fbcf38.png)

**Reply by VorlonCD** [reply]
Search for "JSON response" in the log, but if it is more than 128 chars it truncates it to prevent the log from getting quite as big.

Yes I have to title case everything and convert the percentage to whole number.

See "RESULT" section here:
https://deepstackpython.readthedocs.io/en/latest/objectdetection.html

**Reply by balucanb** [reply]
I started getting this error? I have not changed anything not sure why it is happening- any ideas? It seems to only be on this one camera.  - -   

 Could not upload image C:\Users\baluc\AppData\Local\Temp\RSyardsd.20210426_164408320.jpg to Telegram with chatid '*********': The underlying connection was closed: The connection was closed unexpectedly. [WebException] (In ClsTriggerActionQueue.TelegramUpload:1384 > BotMakeRequestAsync)

**Reply by VorlonCD** [reply]
Does it happen every time?   
Could there have been a lot of alerts right around that time and more than once telegram was trying to be sent?
Temporary internet issue?  
Are connected to VPN?  See if disabling fixes.
If your antivirus has its own firewall, disable.  Wouldnt worry about built in windows firewall.
Reboot?

**Reply by balucanb** [reply]
Might be multiple alerts now that you mention it. I need to check but I think I have a group of cameras triggering anytime any single one of them gets triggered.

**Comment by Yonny24** [comment]
Please can someone tell me where I can find the latest zip for the newest version?
I have version 2.0.1.037

Thanks

**Reply by SankeerthBoddu** [reply]
The latest executable is always here - https://github.com/VorlonCD/bi-aidetection/tree/master/src/UI/Installer ; i was able to install it to the same zip extract I had before. Didn't see any zip locations.
